acme/challenge.go

```diff
@@ -446,158 +441,178 @@ func wireOIDC01Validate(ctx context.Context, ch *Challenge, db DB, jwk *jose.JSO
 	if err != nil {
 		return WrapErrorISE(err, "error unmarshalling challenge data")
 	}
 
-	clientID, err := wire.ParseClientID(challengeValues.ClientID)
+	clientID, err := wire.ParseClientID(wireID.ClientID)
 	if err != nil {
 		return WrapErrorISE(err, "error parsing device id")
 	}
 
+	dpopOptions := prov.GetOptions().GetWireOptions().GetDPOPOptions()
 	issuer, err := dpopOptions.GetTarget(clientID.DeviceID)
 	if err != nil {
 		return WrapErrorISE(err, "invalid Go template registered for 'target'")
 	}
 
-	expiry := strconv.FormatInt(time.Now().Add(time.Hour*24*365).Unix(), 10)
-	cmd := exec.CommandContext( //nolint:gosec // TODO(hs): replace this with Go implementation
-		ctx,
-		dpopOptions.GetValidationExecPath(),
-		"verify-access",
-		"--client-id",
-		challengeValues.ClientID,
-		"--handle",
-		challengeValues.Handle,
-		"--challenge",
-		ch.Token,
-		"--leeway",
-		"360",
-		"--max-expiry",
-		expiry,
-		"--issuer",
-		issuer,
-		"--hash-algorithm",
-		`SHA-256`,
-		"--kid",
-		kid,
-		"--key",
-		file.Name(),
-	)
-
-	stdin, err := cmd.StdinPipe()
+	params := verifyParams{
+		token:      dpopPayload.AccessToken,
+		key:        dpopOptions.GetSigningKey(),
+		accountJWK: accountJWK,
+		issuer:     issuer,
+		wireID:     wireID,
+		challenge:  ch,
+		t:          clock.Now().UTC(),
+	}
+	_, dpop, err := parseAndVerifyWireAccessToken(params)
 	if err != nil {
-		return WrapErrorISE(err, "error getting process stdin")
+		return WrapErrorISE(err, "failed validating token")
 	}
 
-	err = cmd.Start()
-	if err != nil {
-		return WrapErrorISE(err, "error starting validation process")
+	// Update and store the challenge.
+	ch.Status = StatusValid
+	ch.Error = nil
+	ch.ValidatedAt = clock.Now().Format(time.RFC3339)
+
+	if err = db.UpdateChallenge(ctx, ch); err != nil {
+		return WrapErrorISE(err, "error updating challenge")
 	}
 
-	_, err = stdin.Write([]byte(wireChallengePayload.AccessToken))
+	orders, err := db.GetAllOrdersByAccountID(ctx, ch.AccountID)
 	if err != nil {
-		return WrapErrorISE(err, "error writing to stdin")
+		return WrapErrorISE(err, "could not find current order by account id")
+	}
+	if len(orders) == 0 {
+		return NewErrorISE("there are not enough orders for this account for this custom OIDC challenge")
 	}
 
-	err = stdin.Close()
-	if err != nil {
-		return WrapErrorISE(err, "error closing stdin")
+	order := orders[len(orders)-1]
+	if err := db.CreateDpopToken(ctx, order, map[string]any(*dpop)); err != nil {
+		return WrapErrorISE(err, "failed storing DPoP token")
 	}
 
-	err = cmd.Wait()
+	return nil
+}
+
+type cnf struct {
+	Kid string `json:"kid,omitempty"`
+}
+
+type wireAccessToken struct {
+	jose.Claims
+	Challenge  string `json:"chal,omitempty"`
+	Cnf        *cnf   `json:"cnf,omitempty"`
+	Proof      string `json:"proof,omitempty"`
+	ClientID   string `json:"client_id,omitempty"`
+	APIVersion int    `json:"api_version,omitempty"`
+	Scope      string `json:"scope,omitempty"`
+}
+
+type wireDpopToken map[string]any
```

Comment:
Can we use a struct here?

Reply:
After discussion with Wire I think the `map[string]any` suits their use case best.